### ISSUE TITLE
Implement [<StructuralEquality>] and [<StructuralComparison>] for simple class types

### ISSUE THREAD

**Issue by Richiban** [fsharp/fslang-suggestions#554]
# Title of Suggestion

**I propose we** allow [<StructuralEquality>] and [<StructuralComparison>] on simple class types

**The existing way of approaching this problem in F# is** to write the equality / comparison members oneself.

## Pros and Cons 

**The advantages of making this adjustment to F# are** to help reduce the dissonance between classes and records (without requiring us to add more and more class-like features to records).

**The disadvantages of making this adjustment to F# are** just the work involved in implementing such a feature.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL):  M

Related suggestions: #553

## Affidavit (must be submitted)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I would be willing to help implement and/or test this
* [ ] I or my company would be willing to help crowdfund F# Software Foundation members to work on this

**Comment by dsyme** [comment]
Marking approved-in-principle.  My comment from #553: 

@Richiban Possibly the simplest path to this is to allow ``[<StructuralEquality>]`` and ``[<StructuralComparison>]`` on simple class types. 

Specifically, class types where there are no extra captured fields implied by the (method-captured) ``let`` bindings of the class type - or we specify that such extra captured fields are ignored for the purposes of these attributes.

I believe it's actually quite simple to implement this. It does help reduce the dissonance between classes and records (without requiring us to add more and more class-like features to records)

So: 
```
[<StructuralEquality>]
type Qty(value : int) =
    do if value < 0 || value > 99 then invalidArg "value" "That is not a valid quantity"
    member this.Value = value
```
or this (note ``someFunction`` is a method, and no fields are implied, per the F# spec):
```
[<StructuralEquality>]
type Qty(value : int) =
    let someFunction () = do-something ... value ... do-something
    do if value < 0 || value > 99 then invalidArg "value" "That is not a valid quantity"
    member this.Value = value
```
or this (note ``value2`` is local to the construction, and no fields are implied, per the F# spec):
```
[<StructuralEquality>]
type Qty(value : int) =
    let value2 = value % 100
    do if value2 < 0 || value2 > 5 then invalidArg "value" "That is not a valid quantity"
    member this.Value = value
```

The only issue is about the meaning this:
```
[<StructuralEquality>]
type Qty(value : int) =
    let value2 = rand()
    member this.Value = value
    member this.Value2 = value2
```
Does ``value2`` get taken into account for structural equality?  I don't know what people would expect, so we should probably outlaw it, or at least warn that ``value2`` will be ignored for the purposes of structural equality,

Also, there would be a requirement that all arguments to the constructor are type-annotated, rather than type inferred, as for structs today, though I'm sure you're happy with that.

**Comment by Tarmil** [comment]
In your final example I personally would expect `value2` to be taken into consideration; but that is probably because I'm familiar with when a `let` is or isn't turned into a field. A beginner might have different expectations.

It's important to note though that there are also cases like this:

```fsharp
[<StructuralEquality>]
type Qty(value: int) =
    let value2 = if value < 0 then 0 else value
    member this.Value = value2
```

Here, `value` is local to the constructor and `value2` is turned into a field, so the only possible implementation is to use `value2` for structural equality.

**Comment by dsyme** [comment]
> Here, value is local to the constructor and value2 is turned into a field, so the only possible implementation is to use value2 for structural equality.

Yes, good point. However note that if ``StrucutralEquality`` is used and the specification is "it compares arguments" then ``value`` would be captured by the implementation of structural equality, and thus would become a field.
